ggml/src/ggml-vulkan.cpp

```diff
@@ -6441,31 +6464,26 @@ GGML_CALL static ggml_status ggml_backend_vk_graph_compute(ggml_backend_t backen
     // Reserve tensor context space for all nodes
     ctx->tensor_ctxs.resize(cgraph->n_nodes);
 
-    for (int i = 0; i < cgraph->n_nodes; i++) {
-        ggml_vk_build_graph(ctx, cgraph->nodes[i], i, i == last_node, false);
-    }
+    bool first_node_in_batch = true; // true if next node will be first node in a batch
+    int submit_node_idx = 0; // index to first node in a batch
 
+    // submit work every submit_count node to overlap CPU cmdbuffer generation with GPU execution
+    constexpr int submit_count = 50;
     for (int i = 0; i < cgraph->n_nodes; i++) {
-        ggml_tensor * node = cgraph->nodes[i];
-
-        if (ggml_vk_is_empty(node)) {
-            continue;
+        if (first_node_in_batch) {
+            submit_node_idx = i;
         }
 
-        bool ok = ggml_vk_compute_forward(ctx, node, i);
-        if (!ok) {
-            if (node->op == GGML_OP_UNARY) {
-                std::cerr << __func__ << ": error: op not supported UNARY " << node->name << " (" << ggml_unary_op_name(static_cast<ggml_unary_op>(node->op_params[0])) << ")" << std::endl;
-            } else {
-                std::cerr << __func__ << ": error: op not supported " << node->name << " (" << ggml_op_name(node->op) << ")" << std::endl;
-            }
+        // TODO probably it's better to move the submit conter to ggml_vk_build_graph since a lot of nodes might not contain actual vulkan work
+        bool submit = i && ((i % submit_count) == 0);
+        bool enqueued = ggml_vk_build_graph(ctx, cgraph->nodes[i], i, cgraph->nodes[submit_node_idx], submit_node_idx, false, i == last_node, submit);
+
+        if (first_node_in_batch && enqueued) {
+            first_node_in_batch = false;
         }
-#ifdef GGML_VULKAN_CHECK_RESULTS
-        else {
-            ggml_vk_check_results_1(node);
```

Comment:
You have to place this call somewhere else if you remove it, otherwise the operator results checker breaks.

Reply:
Do you have a recommended place where to put it?

I haven't tried to fully understand the check_results_0 / check_results_1 code yet. At a first glance it looks like check_results_0 executes the op on the CPU and check_results_1 downloads the result from the GPU and compares the results.

If that's true I'd put a loop at the end which calls check_results_0 and check_results_1 for each node that returned true and / or add a new check_boolean to the compute_forward_function which executed both checks. Still, a 2nd loop is required to run the check on each node.

Reply:
That is true, yeah. It had to be this way because of inplace tensor operations. The mode forces each Vulkan operator to run in its own command buffer so that it can grab the inputs and run the operation on CPU in the first function, and compare the result in the second function afterwards.

I can take a closer look later if you don't find a good place for it first.

Reply:
With this description it's quite simple to fix the issue moving the check into `ggml_vk_compute_forward` and set `use_fence = true` always if error checking is enabled.

Testing with GGML_VULKAN_CHECKS_ENABLED I found an error as well for `stable-code-instruct-3b-Q8_0.gguf`. At a first glance it looks like a multiply by 2 is missing here.

```
ERROR: avg_err=0.542829 in ADD (check 76)
tensor=000001AB88C68A20 tensor->name=l_out-2 tensor->type: f32 ne0=2560 nb0=4 ne1=2 nb1=10240 ne2=1 nb2=20480 ne3=1 nb3=20480 offset=0
src0=000001AB88C688B0 op=MUL_MAT type=f32 ne0=2560 nb0=4 ne1=2 nb1=10240 ne2=1 nb2=20480 ne3=1 nb3=20480 offset=0
src1=000001AB88C67D30 op=ADD type=f32 ne0=2560 nb0=4 ne1=2 nb1=10240 ne2=1 nb2=20480 ne3=1 nb3=20480 offset=0
First error: result=0.711832 correct=1.41905 i3=0 i2=0 i1=0 i0=1

Result:
               0       1       2       3       4       5       6       7       8       9
      0:    0.05    0.05
      1:    0.71    0.71
      2:    1.39    1.39
      3:   -0.26   -0.26
      4:    0.39    0.39
      5:   -0.64   -0.64
      6:    0.00    0.00
      7:   -0.16   -0.16
      8:   -0.04   -0.04
      9:    0.35    0.35

Correct:
               0       1       2       3       4       5       6       7       8       9
      0:    0.09    0.09
      1:    1.42    1.42
      2:    2.77    2.77
      3:   -0.51   -0.51
      4:    0.78    0.78
      5:   -1.28   -1.28
      6:    0.01    0.01
      7:   -0.32   -0.32
      8:   -0.08   -0.08
      9:    0.69    0.69

Reply:
It does run now, but it only catches very few ops.
```
9 norm-0 op=RMS_NORM avg_err=1.44221e-08
10 v_cache_view-4 (copy of Vcur-4 (transposed)) op=CPY avg_err=0
11 ffn_gate-8 op=MUL_MAT avg_err=3.40353e-06
12 Qcur-13 op=ROPE avg_err=2.74683e-09
13 kqv_merged_cont-17 op=CONT avg_err=0
14 ffn_out-21 op=MUL_MAT avg_err=0.00043354
15 k_cache_view-26 (copy of Kcur-26) op=CPY avg_err=0
16 ffn_norm-30 op=MUL avg_err=0
```
On master it looks like this (just the first layer shown):
```
742 norm-0 op=RMS_NORM avg_err=1.44221e-08
743 attn_norm-0 op=MUL avg_err=0
744 Qcur-0 op=MUL_MAT avg_err=0.00217402
745 Qcur-0 op=ROPE avg_err=4.619e-06
746 Kcur-0 op=MUL_MAT avg_err=0.00230008
747 Kcur-0 op=ROPE avg_err=4.54624e-06
748 Vcur-0 op=MUL_MAT avg_err=0.000835142
749 k_cache_view-0 (copy of Kcur-0) op=CPY avg_err=0
750 v_cache_view-0 (copy of Vcur-0 (transposed)) op=CPY avg_err=0
751 kq-0 op=MUL_MAT avg_err=1.81626e-06
752 kq_soft_max_ext-0 op=SOFT_MAX avg_err=1.65106e-10
753 kqv-0 op=MUL_MAT avg_err=5.40341e-10
754 kqv_merged_cont-0 op=CONT avg_err=0
755 kqv_out-0 op=MUL_MAT avg_err=1.16777e-05
756 ffn_inp-0 op=ADD avg_err=0
757 norm-0 op=RMS_NORM avg_err=1.36306e-08
758 ffn_norm-0 op=MUL avg_err=0
759 ffn_gate-0 op=MUL_MAT avg_err=0.000427647
760 ffn_silu-0 op=UNARY avg_err=1.54837e-09
761 ffn_up-0 op=MUL_MAT avg_err=0.000405687
762 ffn_gate_par-0 op=MUL avg_err=0
763 ffn_out-0 op=MUL_MAT avg_err=3.88136e-05
764 l_out-0 op=ADD avg_err=0
```

Reply:
@mtavenrath I found the issue, here's a possible fix:
```diff
diff --git a/ggml/src/ggml-vulkan.cpp b/ggml/src/ggml-vulkan.cpp
index 411a13c8..81fc6922 100644
--- a/ggml/src/ggml-vulkan.cpp
+++ b/ggml/src/ggml-vulkan.cpp
@@ -6483,9 +6483,11 @@ GGML_CALL static ggml_status ggml_backend_vk_graph_compute(ggml_backend_t backen
         if (enqueued) {
             ++submitted_nodes;

+#ifndef GGML_VULKAN_CHECK_RESULTS
             if (first_node_in_batch) {
                 first_node_in_batch = false;
             }
+#endif
         }

         if (submit) {
```
`first_node_in_batch` has to be true for the checks to work.

Reply:
This shouldn't be required and splatters the check results logic over multiple functions where it'd be great if it is as local as possible.

To get to `ggml_vk_compute_forward` I am setting `last_node_ to true. The main difference between working and non-working is that last_node sets the `exit_node` while submit doesn't. Can you check if changing `last_node` to `submit` solves the issue as well?

```
#if defined(GGML_VULKAN_CHECK_RESULTS) || defined(GGML_VULKAN_PERF)
    // Force context reset on each node so that each tensor ends up in its own context
    // and can be run and compared to its CPU equivalent separately
    last_node = true;
#endif
```

Reply:
It does not. The issue is that it only submits `node_begin`, not `node` and the logic in `ggml_backend_vk_graph_compute` only resets `node_begin` if `first_node_in_batch` is true. My fix is the easiest one, but there's probably a smarter way to make it submit each node separately for the result checker.

Reply:
Okay. I have added your fix to resolve the issue. Thank you for debugging.